File: 152-maximum-product-subarray/maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int sz = nums.size();
        int mini = nums[0], maxi = nums[0], product = nums[0];

        for(int i=1; i<sz; i++){
            int temp = max({nums[i], mini*nums[i], maxi*nums[i]});
            mini = min({nums[i], maxi*nums[i], mini*nums[i]});
            maxi = temp;
            product = max(product, maxi);
        }

        return product;
    }
};

File: 152-maximum-product-subarray/maximum-product-subarray.java
class Solution {
    private int ans = Integer.MIN_VALUE;
    public int maxProduct(int[] nums) {
        int sz = nums.length, tmp = 1;

        solve(nums, sz, nums[0], nums[0], 1);

        return ans;
    }

    private void solve(int nums[], int sz, int pos, int neg, int i){
        ans = Math.max(ans, Math.max(pos, neg));
        if(i == sz) return;

        solve(nums, sz, Math.max(nums[i], Math.max(pos*nums[i], neg*nums[i])), Math.min(nums[i], Math.min(pos*nums[i], neg*nums[i])), i+1);
    }
}